installdir/hlfuzzy.1.in
.TH hldig 1 "2018-04-04"
.\" NAME should be all caps, SECTION should be 1-8, maybe w/ subsection
.\" other parms are allowed: see man(7), man(1)
.SH NAME
hlfuzzy \- fuzzy command-line search utility for the hl://Dig search engine
.SH SYNOPSIS
.B hlfuzzy [-c configfile] [-v] algorithm
.I "[options]"
.SH "DESCRIPTION"
Description to follow here
.SH OPTIONS
.TP
.B \-c \fIconfigfile\fR
Use the specified
.I configfile
instead of the default. 
.TP
.B \-v 
.I Verbose mode
. Use this to get more verbose output while running hlfuzzy. More than one \fIv\fR can be specified (ie. 
.\fI-vv\fR
,
\fI-vvv\fR
 etc
). Specifying more than two v's is probably only useful for debugging purposes.
.SH "Supported Algorithms"
.TP
Different algoritms are supported for searching:
\fIsoundex\fR
Creates  a slightly modified soundex key database. Differences with the standard soundex algorithm are: 
Keys are 6 digits and the first letter is also encoded.
	      
\fImetaphone\fR
Creates a metaphone key database. This algorithm is more specific to English, but  will  get  fewer  "weird"
matches than the soundex algorithm.

\fIaccents\fR
(No description of the algorithm at the moment)

\fIendings\fR
Creates  two  databases  which  can  be  used  to match common word endings. The creation of these databases
requires a list of affix rules and a dictionary which uses those affix rules. The format of the affix  rules
and  dictionary  files are the ones used by the ispell program. Included with the distribution are the affix
rules for English and a fairly small English dictionary.  Other languages can be supported  by  getting  the
appropriate affix rules and dictionaries. These are available for many languages; check the ispell distribu-
tion for more details.

\fIsynonyms\fR
Creates a database of synonyms for words. It reads a text database of synonyms and creates a  database  that
hlsearch  can  then use. Each line of the text database consists of words where the first word will have the
other words on that line as synonyms.

.SH "Notes on searching"
.TP
Please consult specialised literature to find out how the different algorithms work or simply try another 
search stragegy you do not get the intended search results.

.SH "FILES"
.TP
.B @DEFAULT_CONFIG_FILE@
The default configuration file.
.B @DATABASE_DIR@/db.accents.db
(Output) Maps between characters with and without accents for accents fuzzy rule
.B @DATABASE_DIR@/db.metaphone.db
(Output) Database of similar-sounding words for metaphone fuzzy rule
.B @DATABASE_DIR@/db.soundex.db
(Output) Database of similar-sounding words for soundex fuzzy rule
.B @COMMON_DIR@/english.0, @COMMON_DIR@/english.aff
(Input) List of words and affix rules used to generate endings
.B @COMMON_DIR@/root2word.db, @COMMON_DIR@/word2rood.db
(Output) Database used for endings fuzzy rule
.B @COMMON_DIR@/synonyms
(Input) List of groups of words considered synonymous
.B @COMMON_DIR@/synonyms.db
(Output) Database used for synonyms fuzzy rule
.SH "SEE ALSO"
Please refer to the HTML pages (in the hldig-doc package)
.B /usr/share/doc/hldig-doc/html/index.html
and the manual pages
.I hldigconfig(8)
,
.I hlmerge(1)
,
.I hlnotify(1)
,
.I hlsearch(1)
and
.I rundig(1)
for a detailed description of hl://Dig and its commands.
.SH AUTHOR
This manual page was written by Robert Ribnitz. It is based on documentation of hl://Dig, and on earlier manpages provided by Christian Schwarz and Stijn de Bekker.


ChangeLog
2018-09-08  Johnny A. Solbu <johnny@solbu.net>

	* Don't pull external scripts. Serve it from our own server.
	  This also disables any tracking from the CDN provider.

2018-09-07  Johnny A. Solbu <johnny@solbu.net>

	* All tools should now be renamed

2018-09-03  Johnny A. Solbu <johnny@solbu.net>

	* Rename htfuzzy as hlfuzzy
	* Fix compile error on gcc 7, 8, and clang 6
	* Fix various compile warnings
	* Bump version: 0.3.0

2018-08-26  Johnny A. Solbu <johnny@solbu.net>

	* Release version 0.2.0
	* Image files doesn't need execute permission
	* Update TESTING.md with current build requirements
	* Remove non-existant benchmark folder from dist target:
	  git doesn't track empty folders
	* Some Makefile.am fixes


2018-04-04  Johnny A. Solbu <johnny@solbu.net>

	* Remove broken build test link
	* rename ht -> hl in topdir files
	* Complete rename in installdir
	* Autoreconf
	* Apparently several subscripts still use the VERSION file. Should
	fix that sometime
	* Rename mans in configure and make scripts.
	* Rename man files and their contents

2018-04-02  Johnny A. Solbu <johnny@solbu.net>

	* VERSION file is obsolete now that version is in configure.ac
	* Update pot file: copyright are assigned to «the hl://Dig group»

2018-03-27  andy5995 <andy400-dev@yahoo.com>

	* dbbench.cc:fix conversion warning This should fix the warning when `make dist` is run: dbbench.cc:269:20: warning: ISO C++ forbids converting a string
	constant to ‘char*’ [-Wpedantic]    char
	*progname = "dbbench problem...";

2018-03-28  Johnny A. Solbu <johnny@solbu.net>

	* bump version: 0.2.0
	* Add bug address to configure
	* Put version in the configure.ac file, where it belongs
	* Update build tests location
	* Rename url in the last files

2018-03-27  Johnny A. Solbu <johnny@solbu.net>

	* Rename urls in more files
	* Change MSGID_BUGS_ADDRESS, point to the issue tracker
	* Update ChangeLog.md with the new URLs
	* New project management: Rename URLs
	* Don't track compiled translations in revision controll
	* Remove unneeded preceding slash

2018-03-27  andy5995 <andy400-dev@yahoo.com>

	* change htdig.gif to logo.jpg (#56) After changes by @jamessouth, I found a few other areas I forgot to
	mention to him.

2018-03-15  jamessouth <jamesdanielsouth@gmail.com>

	* issue 56 new logo and modified html files

2018-03-15  Andy Alt <andy5995@users.noreply.github.com>

	* Changlog.md:add note about default install paths
	* Makefile.am:remove CVS dist-hook
	* CONTRIBUTING.md:add note about "skip ci"
	* TESTING.md:updated This reflects changes to the new install locations, which was a
	recent patch submitted by @Solbu
	* push regenerated build scripts and makefiles Regenerated due to previous commit
	* configure.ac:change path of www images When installdir/Makefile is run, it uses the value here to prefix
	urls to images in the installdir/html templates with /www/ To be accessible, a symlink can be put in the web root that points
	to that directory.

2018-03-15  Johnny A. Solbu <github@solbu.net>

	* Install things in the proper places (#90)
	  Install things in PREFIX/share/hldir, rather than /var/www

2018-03-12  Andy Alt <andy5995@users.noreply.github.com>

	* Merge pull request #101 from solbu/installer-fixes Installer fixes (Issue #99)
	* republish site from changes in bcdf311 This republishes the site using changes @jamessouth made in
	bcdf3110c78a308404431f3f3a24d33503d283a5 (#81)

2018-03-12  james south <jamessouth@users.noreply.github.com>

	* make sure home page validates (#100) (fixes #81)

2018-03-12  Johnny A. Solbu <johnny@solbu.net>

	* Add ABOUT-NLS to distributed files, indent traling backslash
	* Add missing gettext.h header

2018-03-11  andy5995 <andy400-dev@yahoo.com>

	* ABOUT-NLS:add to distribution
	* remove '#' (forgot to do this in a previous commit)
	* .travis.yml:drop '#' from hldig Freenode channel
	* simplecgen:refactor trim() function
	* list #hldig for the chat room, not ##hldig
	* .travis.yml: change failure and error notifications Only notify on_failure if there's a change Always notify on error
	* .travis.yml:change irc notification options
	* .travis.yml:change osx and Xcode version (#98)
	* .travis.yml:change osx and Xcode version
	* .travis.yml:add matrix closes #76

2018-03-11  Johnny A. Solbu <github@solbu.net>

	* Fix test dir (issue #91) (#97) / Thanks @solbu
	* Fix «make dist» Some tools where renamed at some point but the test commands using
	them where not updated
	* Update test/test_function.in with the renamed filenames

2018-03-09  Andy Alt <andy5995@users.noreply.github.com>

	* Merge pull request #92 from solbu/doc-changes Simplify translation instructions

2018-03-07  andy5995 <andy400-dev@yahoo.com>

	* translation:add ru to LINGUAS, generate gmo file (closes #68)
	* travis.yml:add osx (#76)
	* ne.po:remove empty file

2018-03-07  Svitlana Galianova <svitlana.galianova@gmail.com>

	* added russian translation

2018-03-07  Johnny A. Solbu <github@solbu.net>

	* Uninstall fix (#84) (closes #48)
	* Downgrade required gettext version, to match my system
	* Add Solbu to authors list

2018-03-06  andy5995 <andy400-dev@yahoo.com>

	* ChangeLog.md:note about html docs not being installed anymore

2018-03-06  Johnny A. Solbu <johnny@solbu.net>

	* Run autoreconf
	* Don't install documentaion in docdir RPM packagers prefer to control that dir using a «%doc» macro
	* Ignore .patch files To avoid committing patches by mistake, when one is also working on
	packaging

2018-01-21  andy5995 <andy400-dev@yahoo.com>

	* fix formatting
	* remove Slack, use IRC found some leftovers to remove (see previous commit)
	* remove Slack workspace from contact info IRC should be sufficient for this project.  http://webchat.freenode.net?channels=%23%23hldig&uio=d4

2018-01-15  andy5995 <andy400-dev@yahoo.com>

	* docs/README: explain how to view the html files

2018-01-11  Andy Alt <andy5995@users.noreply.github.com>

	* [#24]installdirs:update search result html, add css (#82)

hlfuzzy/hlfuzzy.cc
//
// hlfuzzy.cc
//
// hlfuzzy: Create one or more ``fuzzy'' indexes into the main word database.
//          These indexes can be used by htsearch to perform a search that uses
//          other algorithms than exact word match.
//
//  This program is meant to be run after htmerge has created the word
//  database.
//
//  For each fuzzy algorithm, there will be a separate database.  Each
//  database is simply a mapping from the fuzzy key to a list of words
//  in the main word database.
//
// Part of the ht://Dig package   <http://www.htdig.org/>
// Copyright (c) 1995-2004 The ht://Dig Group
// For copyright details, see the file COPYING in your distribution
// or the GNU Library General Public License (LGPL) version 2 or later
// <http://www.gnu.org/copyleft/lgpl.html>
//
// $Id: hlfuzzy.cc,v 1.20 2004/05/28 13:15:20 lha Exp $
//
#ifdef HAVE_CONFIG_H
#include "hlconfig.h"
#endif /* HAVE_CONFIG_H */

#include "hlfuzzy.h"
#include "Fuzzy.h"
#include "Accents.h"
#include "Soundex.h"
#include "Endings.h"
#include "Metaphone.h"
#include "Synonym.h"
#include "htString.h"
#include "List.h"
#include "Dictionary.h"
#include "defaults.h"
#include "HtWordList.h"
#include "WordContext.h"
#include "messages.h"

// If we have this, we probably want it.
#ifdef HAVE_GETOPT_H
#include <getopt.h>
#elif HAVE_GETOPT_LOCAL
#include <getopt_local.h>
#endif

int    debug = 0;

void usage();


//*****************************************************************************
// int main(int ac, char **av)
//
int
main(int ac, char **av)
{
    int      c, i;
    extern char  *optarg;
    extern int  optind;
    String    configFile = DEFAULT_CONFIG_FILE;

    //
    // Parse command line arguments
    //
    while ((c = getopt(ac, av, "c:v")) != -1)
    {
  switch (c)
  {
      case 'c':
    configFile = optarg;
    break;

      case 'v':
    debug++;
    break;

      default:
    usage();
  }
    }

  HtConfiguration* config= HtConfiguration::config();
    //
    // Determine what algorithms to use
    //
    List  wordAlgorithms;
    List  noWordAlgorithms;
    for (i = optind; i < ac; i++)
    {
  if (mystrcasecmp(av[i], "soundex") == 0)
  {
      wordAlgorithms.Add(new Soundex(*config));
  }
  else if (mystrcasecmp(av[i], "metaphone") == 0)
  {
      wordAlgorithms.Add(new Metaphone(*config));
  }
  else if (mystrcasecmp(av[i], "accents") == 0)
  {
      wordAlgorithms.Add(new Accents(*config));
  }
  else if (mystrcasecmp(av[i], "endings") == 0)
  {
      noWordAlgorithms.Add(new Endings(*config));
  }
  else if (mystrcasecmp(av[i], "synonyms") == 0)
  {
      noWordAlgorithms.Add(new Synonym(*config));
  }
  else
  {
      reportError(form(_("'%s' is not a supported algorithm"),
           av[i]));
  }
    }
    if (wordAlgorithms.Count() == 0 && noWordAlgorithms.Count() == 0)
    {
  cout << _("hlfuzzy: No algorithms specified\n");
  usage();
    }

    //
    // Find and parse the configuration file.
    //
    config->Defaults(&defaults[0]);
    if (access((char*)configFile, R_OK) < 0)
    {
  reportError(form(_("Unable to find configuration file '%s'"),
       configFile.get()));
    }
    config->Read(configFile);

    // Initialize htword library (key description + wordtype...)
    WordContext::Initialize(*config);

    Fuzzy  *fuzzy;
    if (wordAlgorithms.Count() > 0)
    {
        //
        // Open the word database so that we can grab the words from it.
        //
        HtWordList  worddb(*config);
  if (worddb.Open(config->Find("word_db"), O_RDONLY) == OK)
    {
      //
      // Go through all the words in the database
      //
      List    *words = worddb.Words();
      String    *key;
      Fuzzy    *fuzzy = 0;
      String    word, fuzzyKey;
      int      count = 0;

      words->Start_Get();
      while ((key = (String *) words->Get_Next()))
        {
    word = *key;
    wordAlgorithms.Start_Get();
    while ((fuzzy = (Fuzzy *) wordAlgorithms.Get_Next()))
      {
        fuzzy->addWord(word);
      }
    count++;
    if ((count % 100) == 0 && debug)
      {
        cout << "hlfuzzy: words: " << count << '\n';
        cout.flush();
      }
        }
      if (debug)
        {
    cout << _("hlfuzzy: total words: ") << count << "\n";
    cout << _("hlfuzzy: Writing index files...\n");
        }

      //
      // All the information is now in memory.
      // Write all of it out to the individual databases
      //
      wordAlgorithms.Start_Get();
      while ((fuzzy = (Fuzzy *) wordAlgorithms.Get_Next()))
        {
    fuzzy->writeDB();
        }
      worddb.Close();
      words->Destroy();
      delete words;
      if (fuzzy)
        delete fuzzy;
    }
  else
    {
      reportError(form(_("Unable to open word database %s"), config->Find("word_db").get()));
    }
    }
    if (noWordAlgorithms.Count() > 0)
    {
  noWordAlgorithms.Start_Get();
  while ((fuzzy = (Fuzzy *) noWordAlgorithms.Get_Next()))
  {
      if (debug)
      {
    cout << _("hlfuzzy: Selected algorithm: ") << fuzzy->getName()
         << endl;
      }
      if (fuzzy->createDB(*config) == NOTOK)
        {
    cout << _("hlfuzzy: Could not create database for algorithm: ")
         << fuzzy->getName() << endl;
        }
  }
    }

    if (debug)
    {
  cout << _("hlfuzzy: Done.\n");
    }

    return 0;
}


//*****************************************************************************
// void usage()
//
void
usage()
{
    Usage help;
    cout << _("usage:");
    cout << " hlfuzzy [-c configfile][-v] algorithm ...\n";
    printf (_("This program is part of hl://Dig %s\n\n"), VERSION);
    cout << _("Supported algorithms:\n");
    cout << "\tsoundex\n";
    cout << "\tmetaphone\n";
    cout << "\taccents\n";
    cout << "\tendings\n";
    cout << "\tsynonyms\n";
    cout << "\n";

    cout << _("Options:\n");

    help.config ();
    help.verbose ();

    exit(0);
}


//*****************************************************************************
// void reportError(char *msg)
//
void
reportError(char *msg)
{
    cout << "hlfuzzy: " << msg << "\n\n";
    exit(1);
}


